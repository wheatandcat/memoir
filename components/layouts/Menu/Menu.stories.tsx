import { mockFn } from "@/storyBookUtils/index";
import React from "react";
import Menu from "./Menu";

export default {
  title: "components/layouts/Menu",
};

export const _Menu = () => (
  <Menu
    items={[
      {
        text: "削除",
        color: "error",
        onPress: mockFn("item1"),
      },
      {
        text: "編集",
        color: "secondary",
        onPress: mockFn("item2"),
      },
    ]}
  />
);
