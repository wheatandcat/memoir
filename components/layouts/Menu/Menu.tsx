import { type FC, memo, useState, useCallback, useRef } from 'react';
import {
  Modal,
  StyleSheet,
  type ViewStyle,
  View as RNView,
  TouchableOpacity,
  TouchableWithoutFeedback
} from 'react-native';
import theme from 'config/theme';
import IconButton from '@/components/layouts/IconButton';
import type { FontColor } from 'lib/styledSystem/styleFontColor';
import View from '@/components/elements/View';
import Text from '@/components/elements/Text';
import Divider from '@/components/elements/Divider';

export type Item = {
  text: string;
  color: FontColor;
  testID?: string;
  onPress: (callback?: () => void) => void;
  removeMenu: boolean;
};

export type Props = {
  items: Item[];
};

type State = { viewX: number; viewY: number };

const Menu: FC<Props> = (props) => {
  const [open, setOpen] = useState(false);
  const [key, setKey] = useState<number | null>(null);
  const [state, setState] = useState<State>({ viewX: 0, viewY: 0 });
  const viewRef = useRef<RNView>(null);

  const onLayout = useCallback(() => {
    viewRef.current?.measureInWindow((x, y) => {
      setState({
        viewY: y || 0,
        viewX: x || 0,
      });
    });
  }, []);

  const onPress = useCallback((item: Item, index: number) => {
    if (item.removeMenu) {
      setKey(index);
      setOpen(false);
    } else {
      item.onPress(() => setOpen(false));
    }
  }, []);

  const onModalHide = useCallback(() => {
    if (key === null) {
      return;
    }

    props.items[key].onPress();
  }, [props, key]);

  const style: ViewStyle = {
    top: state.viewY + 20,
    right: 0,
  };

  const onOpen = useCallback(() => {
    setKey(null);
    setOpen(!open);
  }, [open]);

  return (
    <>
      <RNView style={styles.menu} ref={viewRef} onLayout={onLayout}>
        <IconButton
          name="more-horiz"
          size="base"
          onPress={onOpen}
          testID="menu"
        />
        <Modal
          visible={open}
          transparent
          onRequestClose={() => {
            setOpen(!open)
          }}
          animationType="fade"
          onDismiss={onModalHide}
          testID="menu_modal"
        >
          <TouchableWithoutFeedback onPress={() => setOpen(!open)}>
            <View style={[style, styles.menuItem]}>
            {props.items.map((item, index) => (
              <View key={item.text}>
                <TouchableOpacity
                  onPress={() => onPress(item, index)}
                  testID={item.testID}
                >
                  <View p={3}>
                    <Text color={item.color}>{item.text}</Text>
                  </View>
                </TouchableOpacity>
                {index < props.items.length - 1 && <Divider />}
              </View>
            ))}
            </View>
          </TouchableWithoutFeedback>
        </Modal>
      </RNView>
    </>
  );
};

const styles = StyleSheet.create({
  menu: {
    position: 'relative',
    zIndex: 1,
  },
  menuItem: {
    position: 'absolute',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: theme().color.secondary.main,
    backgroundColor: theme().color.background.light,
    width: 200,
    borderRadius: 5,
  },
});

export default memo(Menu);
