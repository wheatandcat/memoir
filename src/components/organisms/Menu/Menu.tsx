import React, { memo, useState, useCallback, useRef } from 'react';
import {
  StyleSheet,
  ViewStyle,
  View as RNView,
  TouchableOpacity,
} from 'react-native';
import theme from 'config/theme';
import IconButton from 'components/molecules/IconButton';
import RNModal from 'react-native-modal';
import { FontColor } from 'lib/styledSystem/styleFontColor';
import View from 'components/atoms/View';
import Text from 'components/atoms/Text';
import Divider from 'components/atoms/Divider';

export type Item = {
  text: string;
  color: FontColor;
  onPress: () => void;
};

type Props = {
  items: Item[];
};

type State = { viewX: number; viewY: number };

const Menu: React.FC<Props> = (props) => {
  const [open, setOpen] = useState(false);
  const [state, setState] = useState<State>({ viewX: 0, viewY: 0 });
  const viewRef = useRef<RNView>(null);

  const onLayout = useCallback(() => {
    viewRef.current?.measureInWindow((x, y) => {
      setState({
        viewY: y || 0,
        viewX: x || 0,
      });
    });
  }, []);

  const onPress = useCallback((item: Item) => {
    setTimeout(() => {
      // Modalは重複して表示できないのでanimationOutTimingを待つ
      item.onPress();
    }, 105);
    setOpen(false);
  }, []);

  const style: ViewStyle = {
    top: state.viewY + 20,
    right: 0,
  };

  return (
    <>
      <RNView style={styles.menu} ref={viewRef} onLayout={onLayout}>
        <IconButton
          name="more-horiz"
          size="base"
          onPress={() => setOpen(!open)}
        />
        <RNModal
          isVisible={open}
          onBackdropPress={() => setOpen(!open)}
          animationIn="fadeInRight"
          animationOut="fadeOutRight"
          backdropOpacity={0}
          animationOutTiming={100}
        >
          <View style={[style, styles.menuItem]}>
            {props.items.map((item, index) => (
              <View key={item.text}>
                <TouchableOpacity onPress={() => onPress(item)}>
                  <View p={3}>
                    <Text color={item.color}>{item.text}</Text>
                  </View>
                </TouchableOpacity>
                {index < props.items.length - 1 && <Divider />}
              </View>
            ))}
          </View>
        </RNModal>
      </RNView>
    </>
  );
};

const styles = StyleSheet.create({
  menu: {
    position: 'relative',
    zIndex: 1,
  },
  menuItem: {
    position: 'absolute',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: theme().color.secondary.main,
    backgroundColor: theme().color.background.light,
    width: 200,
    borderRadius: 5,
  },
});

export default memo(Menu);
