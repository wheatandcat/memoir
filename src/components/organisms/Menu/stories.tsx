import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { mockFn } from 'storyBookUtils/index';
import Menu from './Menu';

storiesOf('organisms/Menu', module).add('Menu', () => (
  <Menu
    items={[
      {
        text: '削除',
        color: 'error',
        onPress: mockFn('item1'),
      },
      {
        text: '編集',
        color: 'secondary',
        onPress: mockFn('item2'),
      },
    ]}
  />
));
