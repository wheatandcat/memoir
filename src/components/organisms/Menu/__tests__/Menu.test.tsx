import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Menu, { Props } from '../Menu';

const propsData = (): Props => ({
  items: [
    {
      text: 'text',
      color: 'primary',
      onPress: jest.fn(),
      removeMenu: false,
    },
  ],
});

describe('components/organisms/Menu/Menu.tsx', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Menu {...propsData()} />);
  });

  it('正常にrenderすること', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
