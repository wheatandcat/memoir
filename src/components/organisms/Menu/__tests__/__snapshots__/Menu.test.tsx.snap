// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/Menu/Menu.tsx 正常にrenderすること 1`] = `
<Fragment>
  <View
    onLayout={[Function]}
    style={
      Object {
        "position": "relative",
        "zIndex": 1,
      }
    }
  >
    <Memo(IconButton)
      name="more-horiz"
      onPress={[Function]}
      size="base"
    />
    <ReactNativeModal
      animationIn="fadeInRight"
      animationInTiming={300}
      animationOut="fadeOutRight"
      animationOutTiming={100}
      avoidKeyboard={false}
      backdropColor="black"
      backdropOpacity={0}
      backdropTransitionInTiming={300}
      backdropTransitionOutTiming={300}
      coverScreen={true}
      customBackdrop={null}
      deviceHeight={null}
      deviceWidth={null}
      hasBackdrop={true}
      hideModalContentWhileAnimating={false}
      isVisible={false}
      onBackButtonPress={[Function]}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalShow={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      panResponderThreshold={4}
      propagateSwipe={false}
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      statusBarTranslucent={false}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
      useNativeDriver={false}
    >
      <View
        style={
          Array [
            Object {
              "right": 0,
              "top": 20,
            },
            Object {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#362740",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "position": "absolute",
              "width": 200,
            },
          ]
        }
      >
        <View
          key="text"
        >
          <TouchableOpacity
            onPress={[Function]}
          >
            <View
              p={3}
            >
              <Text
                color="primary"
                fontFamily="RobotoCondensed-Bold"
                variants="body"
              >
                text
              </Text>
            </View>
          </TouchableOpacity>
        </View>
      </View>
    </ReactNativeModal>
  </View>
</Fragment>
`;
